kaldb/src/main/java/com/slack/kaldb/metadata/core/KaldbPartitioningMetadataStore.java

```diff
@@ -0,0 +1,303 @@
+   *
+   * @see KaldbMetadataStore#KaldbMetadataStore(AsyncCuratorFramework, CreateMode, boolean,
+   *     ModelSerializer, String)
+   */
+  private Watcher buildWatcher() {
+    return event -> {
+      if (event.getType().equals(Watcher.Event.EventType.NodeChildrenChanged)) {
+        curator
+            .getChildren()
+            .forPath(storeFolder)
+            .thenAcceptAsync(
+                (partitions) -> {
+                  // create internal stores foreach partition that do not already exist
+                  partitions.forEach(this::getOrCreateMetadataStore);
+
+                  // remove metadata stores that exist in memory but no longer exist on ZK
+                  Set<String> partitionsToRemove =
+                      Sets.difference(metadataStoreMap.keySet(), Sets.newHashSet(partitions));
+                  partitionsToRemove.forEach(
+                      partition -> {
+                        LOG.info("Closing unused store for partition - {}", partition);
+                        KaldbMetadataStore<T> store = metadataStoreMap.remove(partition);
+                        store.close();
+                      });
+                });
+      }
+    };
+  }
+
+  public CompletionStage<String> createAsync(T metadataNode) {
+    return getOrCreateMetadataStore(metadataNode.getPartition()).createAsync(metadataNode);
+  }
+
+  public void createSync(T metadataNode) {
+    try {
+      createAsync(metadataNode)
+          .toCompletableFuture()
+          .get(DEFAULT_ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new InternalMetadataStoreException("Error creating node " + metadataNode, e);
+    }
+  }
+
+  public CompletionStage<T> getAsync(String partition, String path) {
+    return getOrCreateMetadataStore(partition).getAsync(path);
+  }
+
+  public T getSync(String partition, String path) {
+    try {
+      return getAsync(partition, path)
+          .toCompletableFuture()
+          .get(DEFAULT_ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new InternalMetadataStoreException("Error fetching node at path " + path, e);
+    }
+  }
+
+  /**
+   * Attempts to find the metadata without knowledge of the partition it exists in. Use of this
+   * should be avoided if possible, preferring the getAsync.
+   *
+   * @see KaldbPartitioningMetadataStore#getAsync(String, String)
+   */
+  public CompletionStage<T> findAsync(String path) {
+    return getOrCreateMetadataStore(findPartition(path)).getAsync(path);
+  }
+
+  /**
+   * Attempts to find the metadata without knowledge of the partition it exists in. Use of this
+   * should be avoided if possible, preferring the getSync.
+   *
+   * @see KaldbPartitioningMetadataStore#getSync(String, String)
+   */
+  public T findSync(String path) {
```

Comment:
what do you think of just putting this a a test util instead of exposing it here?

Reply:
This is used in our production code path today as part of the `ReadOnlyChunkImpl.handleChunkAssignment`. In this case we have a `replicaId` which we need to first find, and then upon getting the `replicaId` we try to find the snapshot. In both of these cases we don't know the partition ahead of time.